source/common/network/proxy_protocol.cc

```diff
@@ -61,17 +60,44 @@ void ProxyProtocol::ActiveConnection::onReadWorker() {
     return;
   }
 
-  if (proxy_line.find(PROXY_TCP4) != 0) {
+  // Parse proxy protocol line with format: PROXY TCP4/TCP6 SOURCE_ADDRESS DESTINATION_ADDRESS
+  // SOURCE_PORT DESTINATION_PORT.
+  auto line_parts = StringUtil::splitKeep(proxy_line, " ", true);
+
+  if (line_parts.size() != 6 || line_parts[0] != "PROXY") {
     throw EnvoyException("failed to read proxy protocol");
   }
 
-  size_t index = proxy_line.find(" ", PROXY_TCP4.size());
-  if (index == std::string::npos) {
+  Address::IpVersion protocol_version;
+  if (line_parts[1] == "TCP4") {
+    protocol_version = Address::IpVersion::v4;
+  } else if (line_parts[1] == "TCP6") {
+    protocol_version = Address::IpVersion::v6;
+  } else {
     throw EnvoyException("failed to read proxy protocol");
   }
 
-  size_t addr_len = index - PROXY_TCP4.size();
-  std::string remote_address = proxy_line.substr(PROXY_TCP4.size(), addr_len);
+  // Error check the source and destination fields. Currently the source port, destination address,
+  // and destination port fields are ignored. Remote address refers to the source address.
+  Address::InstanceConstSharedPtr remote_address = Utility::parseInternetAddress(line_parts[2]);
+  Address::InstanceConstSharedPtr destination_address =
+      Utility::parseInternetAddress(line_parts[3]);
+  auto remote_version = remote_address->ip()->version();
+  auto destination_version = destination_address->ip()->version();
+  if (remote_version != protocol_version || destination_version != protocol_version ||
+      remote_version != destination_version) {
+    throw EnvoyException("failed to read proxy protocol");
+  }
+
+  uint64_t remote_port, destination_port;
+  try {
+    remote_port = std::stol(line_parts[4], nullptr);
+    destination_port = std::stol(line_parts[5], nullptr);
+  } catch (const std::invalid_argument& ia) {
```

Comment:
In general, I'm not sure how aggressive to be in error checking these fields because they aren't used. For example, should we also check whether the value is within the port range?